### PR TITLE
Add day picker to Agenda and Day view

### LINKED ISSUE #3508: Calendar navigation improvements
As a user of the Tutanota calendar i want to have a view selector visible on each calendar view so that i can easily switch between the views and don't have to deal with back and forth navigation to see events of one day.

## AC  
* There is an option to switch between views in the toolbar  
   * 1 column layout: (Agenda, Day, Month)
   * 2 column layout: (Agenda, Day, Week, Month)
   * 3 column layout: (Agenda, Day, Week, Month)
* A day selector appears for day and agenda view
    *  First row shows weekday name. Week starts on monday or sunday depending on the setting.
    * Day of the month number appears below the week day name.
    * A dot should appear below the day of the month number to indicate that there is at least one event on that day
    * Left/right swipe gesture changes the week
    * When the week changes the selected day changes to the same week day of that week.
* Agenda view shows only the events of the selected day
   * First row shows event title
   * Second row shows start and end time
   * Second row shows "all-day"/ "Ganztägig" for all day events
   * For long/multi day events it shows 
       * start time if the event starts on that day and is not an all day event
       * end time if the event  ends on that day and is not an all day event
       * "all-day"/ "Ganztägig" in all other cases
* There is a "Today" selector in the toolbar which always sets the selected day to the current day. The icon shows the current day number.
* There is a button to extend the day selector to a day of month selector
* Selecting a day in the month view opens the agenda view for that day.

## Images

calendar-navigation

## Tasks
- [x] Add option to switch between views in toolbar
   - [x] Single column layout as shown in picture above
   - [ ] Today selector with icon showing current day number
- [x] #6099
- [x] Day selector can be extended to month selector
   - [x] Selecting day in the month opens that day
   - [x] Single dot indicator below days with at least one event 
- [x] Agenda view
  - [x] Entry with First row being event title and second row showing time
  - [x] "All day" for all day events (make sure translations work)
- [x]  Selecting a day in the month view opens the agenda view for that day (instead of day view)

## ui team review changes
- [x] Different color for padding days and header, try lighter grey (Month View - 1/2C Layout)
- [x] Make text for week day names bold, always use one row for week day name and one row for the day in all views.
- [x] Use only one letter for week day name in mobile layout, for two and three column layout use three letter. 
- [x] Reduce the distance between week day names and days
- [x] Change icon for  "today" action so that it looks like the proposed one. Ensure that the size is the same as the week selector.
- [x] Only show month name next to the hamburger menu in each view.
- [x] Use pop up as month day selector on larger screens (2 column layout)
- [x] Change selected/current day indication
   - [x] Use light red background for selected days
   - [x] Try grey background for current day
   - [x] Remove bar indicator from all views
   - [x] Only show selected day in day and agenda view 
- [x] Try to remove line below the header in time based views.
- [x] Allow week view for 1 column layout.
- [x] Make smaller time column. Use only two letters for time. am/pm should be rendered below
- [x] Shift day selector to the right by the size of the time column, so week day names align with week columns
- [x] Apply same shifting for the day selector in day/week and agenda view.

## Review 22.11
- [x] Highlight today on Day Selector in all views
- [x] On Week/Day view, always display long events for the current week and size them according to the size of the days in Day Selector
- [x] Fix the Day Selector expanded when the user turns the device
- [x] Fix the Week day names on Month view
- [x] Highlight the entire selected week in the Day Selector when expanded on the week view

## Review 24.11
- [x] Month view padding days:
  - [x] Try lighter grey for day numbers.
  - [x] Keep white background.
  - [x] Remove line below header.
  - [x] Change border color to #e1e1e1
- [x] Day /Agenda
  - [x] Not use different color for days when the day selector just shows one week.
  - [x] Change opacity for selected day indicator to 0.20
  - [x] Agenda. Align events to content week grid
  - [x] Agenda (two column layout): fix padding-left for day selector.
- [x] Day view
  - [x] Show only long/all day events of that day.
- [x] View selector
  - [x] Change order of items to Agenda, Day, Week, Month.
- [x] Week view
  - [x] Use rounded corners for week selection in expanded day selector
- [x] Fix today button tooltip

## Notes

| Problem | Platform | Suggestions |
|--------|--------|--------|
|When in day view i cannot easily switch to another day, i have to go back/select to the month view and then pick another day.| mobile | 1. Allow opening a date picker from the day view, 2. Show days of the week on top of the day view.|
| Week view for mobile devices. Possible implementations: | mobile | 1. Put day numbers at top of day view that allow for selecting a specific day 2.  A week view that looks like a small version of month view (3x3 or 2x4) 3. Show existing week view on larger mobile devices (two column layout)|
| Switch between views should be accessible without opening the side menu so it is clear from the user interface how to change views. Example: when i am in the day view on mobile i have to go back to the month view, then open the side menu and then select agenda to have an agenda style view. | mobile | Add view switcher to the toolbar. |
| On mobile devices the month view opens by default and i have to navigate to the day view by clicking on the tile. | mobile | Show day view by default, introduce date picker, show days of the week on top of the day view. |
| In the month view i often cannot see all events of the current day because they don't fit into one tile, more events are marked with (+x). | mobile/desktop | 1. Desktop: Introduce third column which shows a day view of the selected day. 2.Mobile Do not navigate to day view but rather show larger tile/dialog with all the events on that day. 3. Navigate to agenda view| 
| In the day view we show a time based calendar. To get an overview of what is happening on that day it would be better to have agenda style view.  | mobile | 1. Allow switching between agenda and time based calendar from day view 2.Reduce the size for one hour so more hours fit on screen 3. Scroll to the first event of that day. | 
| In desktop layout i have to select the day or week view from the nav column explicitly to see all entries of that day. | desktop | Introduce third column in desktop layout which shows a day view of the selected day. | 

## Test Notes
- [ ] A day picker is shown in Day view
- [ ] A day picker is shown in Agenda view
- [ ] The user can expand/contract the day picker
- [ ] The day picker is only shown in single column layout
- [ ] The other layouts doesn't appears
- [ ] Days if events have a dot bellow the day number
- [ ] Adding an event will add a dot to the respective day
- [ ] All the functions of the agenda and day view still working
- [ ] The week in the day picker starts according to user's config
- [ ] Swiping the day picker while expanded:
   - [ ] To the right, goes to the previous month
   - [ ] To the left, goes to the next month
   - [ ] The selected day should be the same day of month
 - [ ] Swiping the day picker while contracted:
   - [ ] To the right, goes to the previous week
   - [ ] To the left, goes to the next week
   - [ ] The selected day should be the same day of week
 - [ ] View type selector is shown in the leftmost column on desktop
 - [ ] View type selector is show in the header on mobile and it works
 - [ ] Today button works is displayed in header on mobile and it works
 - [ ] Clicking on a day (anywhere, including day number) in month view on mobile opens agenda for that day
   - [ ] it still creates an event on desktop when clicking not on  day number
   - [ ] It still opens day view on desktop if you click on the day number
 - [ ] On desktop agenda always shows  events of the next days, independent of the selected date
 - [ ] Mobile agenda only displays the part of the event that falls on the day
 - [ ] Mobile & desktop agenda shows "all day" correctly

### LINKED ISSUE #5778: agenda dates changing with same events
Description:
The agenda dates are changing based on what date in the calendar you are on. 
This is confusing because the events in the agenda are always the upcoming 2 weeks.

1. Go to Calendar
2. Select any date 
3. Go to Agenda

Expected:
The Agenda should always display the current date + 2 weeks on top.

This issue happens on Desktop/Web & Mobile
Tutanota Ver.: 3.116.8

## Test Notes
on calendar navigation improvement issue

### LINKED ISSUE #6099: Add day selector for day and agenda view
## AC

A day selector appears for day and agenda view

- First row shows weekday name. 
- Week starts on monday or sunday depending on the setting.
- Day of the month number appears below the week day name.
- A dot should appear below the day of the month number to indicate that there is at least one event on that day
- Left/right swipe gesture changes the week
- When the week changes the selected day changes to the same week day of that week.

## Tasks

- [x] Show weekday name on first row
- [x] Week starts on correct day (this is a user setting)
- [x] swiping left or right changes the week (look into possibly adding buttons, so the week can be changed with a click)
- [x] When the week changes, the selected day changes to the same day of the week
- [x] Single dot indicator below days with at least one event 
- [x] Day selector can be extended to month selector
- [x] Smaller circle for the selected day
- [x] Try to adapt colors closer to the mockup
- [ ] Adapt the weekdays background in month and week view and use just one char

## Test Notes:
these are on the calendar navigation improvement issue

### PR DESCRIPTION
**Pull request by murilopereirame**
Add Week/Month preview for previous and next month in Agenda and Day
views.

Add a vertical arrow for showing whether the day selector
is displaying by month or week to the calendar header in the agenda view
on mobile.

Change calendar navigation on mobile, change agenda style

All view types are on the same level now.

View type switcher and today button are in the header.

Month view navigates to agenda.

Mobile agenda only displays events for a single day

If the date from URL is different from today (in desktop) and current
view is Agenda, we move the url to today and show the right information

Co-Authors: @rezbyte @charlag 

close #6099 
close #3508
fix #5778

### CODE REVIEW

**Comment by wrdhub** [comment]
Here are some things I noticed while smoke testing this:
- swiping on day view changes day, but not day indicator in selector
- swiping left and right on the day picker should show a preview of the next days/month.
- In addition, I think only numbers should shift and day of week letter should not shift (may just be my personal preference though)

File: src/calendar/view/CalendarAgendaView.ts [hunk 1]
```diff
@@ -28,38 +32,92 @@ type Attrs = {
...
+
+	private handleDayChange = (selectedDate: Date) => {
+		this.selectedDate = selectedDate
+		this.today = getStartOfDayWithZone(this.selectedDate, this.zone)
+		this.days = getNextFourteenDays(this.today)
+
+		this.onDateSelected(selectedDate)
+	}
+
+	private handleCalendarSwipe = (isNext: boolean) => {
```
Comment (wrdhub) [inline src/calendar/view/CalendarAgendaView.ts:65]: This is for the mini day-selector calendar right? There is another function named exactly the same for "main" calendar, we may want to name this differently to differentiate.

File: src/calendar/view/CalendarView.ts [hunk 2]
```diff
@@ -810,3 +835,49 @@ export class CalendarView extends BaseTopLevelView implements TopLevelView<Calen
 		new CalendarEventPopup(popupModel, rect, htmlSanitizer).show()
 	}
 }
+
+export class CalendarSwipeHandler extends SwipeHandler {
```
Comment (wrdhub) [inline src/calendar/view/CalendarView.ts:839]: I would put this in its own file.

File: src/gui/date/DatePicker.ts [hunk 3]
```diff
@@ -317,7 +395,7 @@ export class VisualDatePicker implements Component<VisualDatePickerAttrs> {
 
 	private renderWeek(week: Array<CalendarDay>, attrs: VisualDatePickerAttrs): Children {
 		return m(
-			".flex.flex-space-between",
+			`.flex.flex-space-between`,
```
Comment (wrdhub) [inline src/gui/date/DatePicker.ts:398]: Leave the double quotes

File: src/gui/date/DatePicker.ts [hunk 4]
```diff
@@ -414,20 +422,23 @@ export class VisualDatePicker implements Component<VisualDatePickerAttrs> {
...
-		const size = px(this.getElementWidth(attrs))
 		const eventForDay = attrs.eventsForDays?.get(date.getTime())
 		const hasEvent = eventForDay && eventForDay.length > 0
 		const isExpandableDatePicker = attrs.isExpandableDatePicker ?? false
+		const size = this.getElementWidth(attrs)
 		const isSelectedDay = isSameDayOfDate(date, attrs.selectedDate)
+		const dateIndicator =
+			isPaddingDay && !isExpandableDatePicker
+				? ""
+				: getDateIndicator(date, attrs.selectedDate).replace("accent-bg", isExpandableDatePicker ? "accent-bg-translucent" : ".accent-bg")
```
Comment (charlag) [inline src/gui/date/DatePicker.ts:433]: why would you replace the selector?

not only string manipulation on selectors is extremely dangerous (it's a bit safer using `class` prop) but also in this case really pointless, just inline this function

File: src/gui/date/DatePicker.ts [hunk 5]
```diff
@@ -342,16 +342,24 @@ export class VisualDatePicker implements Component<VisualDatePickerAttrs> {
...
 
-	private renderCarouselPage(isExpanded: boolean, currentDate: Date, attrs: VisualDatePickerAttrs) {
-		const calendarMonth = getCalendarMonth(currentDate, attrs.startOfTheWeekOffset, true)
+	private renderCarouselPage(isExpanded: boolean, date: Date, attrs: VisualDatePickerAttrs, position: SwipeCalendarOrder) {
+		let calendarMonth = getCalendarMonth(date, attrs.startOfTheWeekOffset, true)
 		const weeks = calendarMonth.weeks
-		const selectedWeek = assertNotNull(weeks.find((w) => w.some((calendarDay) => currentDate.getTime() === calendarDay.date.getTime())))
+		const selectedWeek = assertNotNull(weeks.find((w) => w.some((calendarDay) => date.getTime() === calendarDay.date.getTime())))
+
+		// We need to recalculate the month since when the previous/next week
```
[the remark this answers is not in the document]
Comment (charlag) [reply src/gui/date/DatePicker.ts:356]: another thing I've realized is that we always render month, even when it's not expanded, even for other pages. It's not great for performance I imagine.

**Comment by charlag** [comment]
There's one more (non-critical but annoying) issue I've noticed

[Screencast from 13.11.2023 17:24:44.webm](https://github.com/tutao/tutanota/assets/3099142/313c8300-486c-4245-bb90-2d0c74babbb1)

File: src/gui/date/DatePicker.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (charlag) [reply src/gui/date/DatePicker.ts:356]: okay I understand the code now but it's very confusing

**Comment by charlag** [comment]
I pushed fixed for some issues but some remain still:
 - like in the video above, the space between months is less than between days in the month. It should be half the space between the days, currently there's no space (replace `space-between` with `space-around` if you want)
 - there's a certain threshold for swipe handler. I would argue that it is too big (maybe also true for PageView). It makes it feel kinda janky
 -  - vertical space below the weekdays is too big
 - We should add previous/next buttons to the header if it's 2-column layout

**Comment by wrdhub** [comment]
I think things look really good, the only issue I have found is that there is this odd flickering when expanding the day selector. It only happens on day view (in agenda view it is very smooth), and I saw this on our testing device and mine (both android, for the record).

File: src/calendar/date/DaySelectorPopup.ts [hunk 6]
```diff
@@ -131,15 +131,30 @@ export class DaySelectorPopup implements ModalComponent {
 		this.currentDate = incrementMonth(this.currentDate, forward ? 1 : -1)
 	}
 
+	// Sets the content div (.main-view) to inert, disabling the ability to be focused, this traps the
+	// focus to the popup, releasing it just when the popup is closed and the inert property removed.
+	private focusHandling(trap: boolean) {
```
Comment (charlag) [inline src/calendar/date/DaySelectorPopup.ts:136]: should we call it `enableTrapFocus` or something?

Comment (murilopereirame) [reply src/calendar/date/DaySelectorPopup.ts:136]: > should we call it `enableTrapFocus` or something?

"enable" makes me think that it only enables it. Maybe ```turnTrapFocus(on: boolean)```?

Comment (charlag) [reply src/calendar/date/DaySelectorPopup.ts:136]: yeah sounds better!

File: src/calendar/view/CalendarAgendaView.ts [hunk 7]
```diff
@@ -103,7 +105,11 @@ export class CalendarAgendaView implements Component<Attrs> {
 				color: theme.list_message_bg,
 			})
 		} else {
-			return m(".pt-s.flex.mlr.mb-s.col", this.renderEventsForDay(events, attrs.selectedDate, getTimeZone(), attrs.groupColors, attrs.onEventClicked))
+			return m(
+				".pt-s.flex.mlr.mb-s.col",
+				{ style: { ...(!styles.isDesktopLayout() ? { marginLeft: px(timeWidth) } : {}) } },
```
[the remark this answers is not in the document]
Comment (murilopereirame) [reply src/calendar/view/CalendarAgendaView.ts:110]: Yes, my mistake, I'll fix it

**Comment by murilopereirame** [comment]
> Screenshot from 2023-11-27 11-35-23
> 
> text color is wrong for today button, otherwise looks good?
> 
> View picker order on desktop should be the same as on mobile IMO, I think it's okay to already change that

Is better to use black or the accent color to today button? I'm afraid that black doesn't have enough contrast

**Pull request merged**